Analysis/WWCategorisation/src/WWCategorisationProcessor.cc

```diff
@@ -0,0 +1,436 @@
+          general_mcdaughters.push_back(daughters[i]); //including mc information of remaining daughters that are not from W or Z
+          general_daughters_pdgs.push_back(daughters_pdg);  //including PDGs of remaining daughters that are not from W or Z 
+        } //by the end we get all 4f daughters PDGs and all 4f daughters mc info
+      }
+    }
+    else if(daughters.size() == 4){ //go through all daughters of e-/e+ and save their PDG, then use the function again to count and categorise the events
+      for(unsigned int i = 0; i < daughters.size(); ++i){
+        general_mcdaughters.push_back(daughters[i]);
+        general_daughters_pdgs.push_back(abs(daughters[i]->getPDG()));
+      }
+    }else{ //PRINT OUT ACTUAL SIZE IF NOT ANY OF THE ONES MENTIONED
+      streamlog_out(DEBUG) << "actual size of daughters: " << daughters.size() << std::endl;
+    }
+
+    int channel = CountingLeptons(general_daughters_pdgs); //which channel -> based of how many final state leptons we found
+    if (channel == 0) true_cat = 0; // hadronic
+    else if (channel == 2) true_cat = 5; // leptonic
+    else if (channel == 1){ // semileptonic, do sub-categorisation
+      streamlog_out(DEBUG) << "MC daughters: [ ";
+      for (unsigned int k = 0; k < general_daughters_pdgs.size(); ++k){
+        streamlog_out(DEBUG) << general_daughters_pdgs[k] << ", ";
+
+        if (IsLepton(general_daughters_pdgs[k])){
+        true_cat = _sl_subcat[general_daughters_pdgs[k]];
+        if (general_daughters_pdgs[k] == 11) myelectron = general_mcdaughters[k];
+        break;
+        }
+      }
+    }
+    streamlog_out(DEBUG) << "]" << std::endl;
+    streamlog_out(DEBUG) << "How many leptons: " << channel << std::endl;
+  }
+  else{ // particle 5 is not e+/-, print for debugging
+      streamlog_out(MESSAGE) << "MCParticle no. 5 PDG: " << incoming_pdg << std::endl;
+  }
+
+  // very forward (i.e. invisible) electrons in semileptonic channel from 'single-W' contribution go to separate true category
+  if (true_cat == 2){
+    std::vector<MCParticle*> deepest_electron{}; // sl electron after FSR -> getting daughters of the electron until we have a FSR electron (radiated a photon)
+    bool last = false;
+    while(!last){
+      deepest_electron = myelectron->getDaughters();
+      last = true;
+      for(unsigned int w = 0; w < deepest_electron.size(); w++){
+        int deepest_electron_pdg = abs(deepest_electron[w]->getPDG());
+        if(((deepest_electron_pdg > 90) && (deepest_electron_pdg < 99)) || (deepest_electron_pdg == 11)){
+          myelectron = deepest_electron[w];
+          last = false;
+          break;
+        }
+      }
+    }
+    TLorentzVector lv = TLorentzVector(myelectron->getMomentum(), myelectron->getEnergy());
+    double mctheta = lv.Theta();
+    double cosine_mctheta = fabs(cos(mctheta));
+    streamlog_out(DEBUG) << "cosine(#theta): " << cosine_mctheta << std::endl;
+    if(cosine_mctheta > 0.994) true_cat = 1;
```

Comment:
Can you make this configurable?

Reply:
You mean the cos theta cut value?

Reply:
yes

Reply:
Implemented, including dynamic labeling of the confusion matrix plot.
Quick check with different values made to validate content-wise.